qiskit/circuit/library/phase_estimation.py

```diff
@@ -97,3 +102,70 @@ def __init__(
 
         super().__init__(*circuit.qregs, name=circuit.name)
         self.compose(circuit.to_gate(), qubits=self.qubits, inplace=True)
+
+
+def phase_estimation(
+    num_evaluation_qubits: int,
+    unitary: QuantumCircuit,
+    name: str = "QPE",
+) -> QuantumCircuit:
+    r"""Phase Estimation circuit.
+
+    In the Quantum Phase Estimation (QPE) algorithm [1, 2, 3], the Phase Estimation circuit is used
+    to estimate the phase :math:`\phi` of an eigenvalue :math:`e^{2\pi i\phi}` of a unitary operator
+    :math:`U`, provided with the corresponding eigenstate :math:`|\psi\rangle`.
+    That is
+
+    .. math::
+
+        U|\psi\rangle = e^{2\pi i\phi} |\psi\rangle
+
+    This estimation (and thereby this circuit) is a central routine to several well-known
+    algorithms, such as Shor's algorithm or Quantum Amplitude Estimation.
+
+    Args:
+        num_evaluation_qubits: The number of evaluation qubits.
+        unitary: The unitary operation :math:`U` which will be repeated and controlled.
+        name: The name of the circuit.
+
+    **Reference Circuit:**
+
+    .. plot::
+
+       from qiskit.circuit import QuantumCircuit
+       from qiskit.circuit.library import phase_estimation
+       unitary = QuantumCircuit(2)
+       unitary.x(0)
+       unitary.y(1)
+       circuit = phase_estimation(3, unitary)
+       circuit.draw('mpl')
+
+    **References:**
+
+    [1]: Kitaev, A. Y. (1995). Quantum measurements and the Abelian Stabilizer Problem. 1–22.
+        `quant-ph/9511026 <http://arxiv.org/abs/quant-ph/9511026>`_
+
+    [2]: Michael A. Nielsen and Isaac L. Chuang. 2011.
+         Quantum Computation and Quantum Information: 10th Anniversary Edition (10th ed.).
+         Cambridge University Press, New York, NY, USA.
+
+    [3]: Qiskit
+        `textbook <https://github.com/Qiskit/textbook/blob/main/notebooks/ch-algorithms/
+        quantum-phase-estimation.ipynb>`_
+
+    """
+    qr_eval = QuantumRegister(num_evaluation_qubits, "eval")
+    qr_state = QuantumRegister(unitary.num_qubits, "q")
+    circuit = QuantumCircuit(qr_eval, qr_state, name=name)
+
+    circuit.h(qr_eval)  # hadamards on evaluation qubits
+
+    for j in range(num_evaluation_qubits):  # controlled powers
+        circuit.compose(unitary.power(2**j).control(), qubits=[j] + qr_state[:], inplace=True)
+
+    circuit.append(QFTGate(num_evaluation_qubits).inverse(), qr_eval[:])
+
+    reversal_pattern = list(reversed(range(num_evaluation_qubits)))
```

Comment:
Could you verify that transpiling the circuit will correctly remove all Swaps? 🙂

Reply:
Yes, I have added the test.